File: scanner.go
package main

import "strconv"

type TokenType int

const (
	LeftParen TokenType = iota
	RightParen
	LeftBrace
	RightBrace
	Comma
	Dot
	Minus
	Plus
	Semicolon
	Slash
	Star
	Bang
	BangEqual
	Equal
	EqualEqual
	Greater
	GreaterEqual
	Less
	LessEqual
	Identifier
	String
	Number
	And
	Class
	Else
	False
	Fun
	For
	If
	Nil
	Or
	Print
	Return
	Super
	This
	True
	Var
	While
	Eof
)

type Token struct {
	tokenType TokenType
	lexeme    string
	literal   any
	line      int
}

type Scanner struct {
	source  string
	tokens  []Token
	start   int
	current int
	line    int
}

func newScanner(source string) *Scanner {
	return &Scanner{
		source: source,
		line:   1,
	}
}

func (s *Scanner) scanTokens() []Token {
	for !s.isAtEnd() {
		s.start = s.current
		s.scanToken()
	}

	s.tokens = append(s.tokens, Token{
		tokenType: Eof,
		lexeme:    "",
		line:      s.line,
	})
	return s.tokens
}

func (s *Scanner) scanToken() {
	c := s.advance()
	switch c {
	case '(':
		s.addToken(LeftParen, nil)
	case ')':
		s.addToken(RightParen, nil)
	case '{':
		s.addToken(LeftBrace, nil)
	case '}':
		s.addToken(RightBrace, nil)
	case ',':
		s.addToken(Comma, nil)
	case '.':
		s.addToken(Dot, nil)
	case '-':
		s.addToken(Minus, nil)
	case '+':
		s.addToken(Plus, nil)
	case ';':
		s.addToken(Semicolon, nil)
	case '*':
		s.addToken(Star, nil)
	case '!':
		if s.match('=') {
			s.addToken(BangEqual, nil)
		} else {
			s.addToken(Bang, nil)
		}
	case '=':
		if s.match('=') {
			s.addToken(EqualEqual, nil)
		} else {
			s.addToken(Equal, nil)
		}
	case '<':
		if s.match('=') {
			s.addToken(LessEqual, nil)
		} else {
			s.addToken(Less, nil)
		}
	case '>':
		if s.match('=') {
			s.addToken(GreaterEqual, nil)
		} else {
			s.addToken(Greater, nil)
		}
	case '/':
		if s.match('/') {
			for s.peek() != '\n' && !s.isAtEnd() {
				s.advance()
			}
		} else if s.match('*') {
			for s.peek() != '*' && s.peekNext() != '/' && !s.isAtEnd() {
				if s.peek() == '\n' {
					s.line++
				}
			}

			if s.isAtEnd() {
				lumi.scanError(s.line, "unclosed block comment")
			}
		} else {
			s.addToken(Slash, nil)
		}
	case ' ':
	case '\t':
	case '\r':
	case '\n':
		s.line++
	case '"':
		s.string()
	default:
		if isDigit(c) {
			s.number()
			return
		} else if isAlpha(c) {
			s.identifier()
		} else {
			lumi.scanError(s.line, "Unexpected Character")
		}
	}
}

func (s *Scanner) addToken(t TokenType, literal any) {
	s.tokens = append(s.tokens, Token{
		tokenType: t,
		lexeme:    s.source[s.start:s.current],
		literal:   literal,
		line:      s.line,
	})
}

var keywords = map[string]TokenType{
	"and":    And,
	"class":  Class,
	"else":   Else,
	"false":  False,
	"for":    For,
	"fun":    Fun,
	"if":     If,
	"nil":    Nil,
	"or":     Or,
	"print":  Print,
	"return": Return,
	"super":  Super,
	"this":   This,
	"true":   True,
	"var":    Var,
	"while":  While,
}

func (s *Scanner) identifier() {
	for isAlphaNumeric(s.peek()) {
		s.advance()
	}
	lexeme := s.source[s.start:s.current]

	typ, ok := keywords[lexeme]
	if !ok {
		typ = Identifier
	}
	s.addToken(typ, lexeme)
}

func (s *Scanner) number() {
	for isDigit(s.peek()) {
		s.advance()
	}

	if s.peek() == '.' && isDigit(s.peekNext()) {
		s.advance()

		for isDigit(s.peek()) {
			s.advance()
		}
	}

	number, err := strconv.ParseFloat(s.source[s.start:s.current], 64)
	if err != nil {
		lumi.scanError(s.line, "invalid number literal")
		return
	}

	s.addToken(Number, number)
}

func (s *Scanner) string() {
	for s.peek() != '"' && !s.isAtEnd() {
		if s.peek() == '\n' {
			s.line++
		}
		s.advance()
	}

	if s.isAtEnd() {
		lumi.scanError(s.line, "Unterminated string")
		return
	}

	s.advance()

	lexeme := s.source[s.start+1 : s.current-1]
	s.addToken(String, lexeme)
}

func (s *Scanner) peekNext() byte {
	if (s.current + 1) >= len(s.source) {
		return 0
	}
	return s.source[s.current+1]
}

func (s *Scanner) peek() byte {
	if s.isAtEnd() {
		return 0
	}
	return s.source[s.current]
}

func (s *Scanner) match(ch byte) bool {
	if s.isAtEnd() {
		return false
	}
	if !(s.source[s.current] == ch) {
		return false
	}

	s.current++
	return true
}

func (s *Scanner) advance() byte {
	ch := s.source[s.current]
	s.current++
	return ch
}

func (s *Scanner) isAtEnd() bool {
	return s.current >= len(s.source)
}

func isAlphaNumeric(ch byte) bool {
	return isAlpha(ch) || isDigit(ch)
}

func isAlpha(ch byte) bool {
	return (ch >= 'a' && ch <= 'z') || (ch >= 'A' && ch <= 'Z') || ch == '_'
}
func isDigit(ch byte) bool {
	return ch >= '0' && ch <= '9'
}

File: scanner.py
from enum import Enum, auto
from typing import Any
from error import error

class TokenKind(Enum):
    # Single-character tokens
    LEFT_PAREN = auto()
    RIGHT_PAREN = auto()
    LEFT_BRACE = auto()
    RIGHT_BRACE = auto()
    COMMA = auto()
    DOT = auto()
    MINUS = auto()
    PLUS = auto()
    SEMICOLON = auto()
    SLASH = auto()
    STAR = auto()

    # One or two character tokens
    BANG = auto()
    BANG_EQUAL = auto()
    EQUAL = auto()
    EQUAL_EQUAL = auto()
    GREATER = auto()
    GREATER_EQUAL = auto()
    LESS = auto()
    LESS_EQUAL = auto()

    # Literals
    IDENTIFIER = auto()
    STRING = auto()
    NUMBER = auto()

    # Keywords
    AND = auto()
    CLASS = auto()
    ELSE = auto()
    FALSE = auto()
    FUN = auto()
    FOR = auto()
    IF = auto()
    NIL = auto()
    OR = auto()
    PRINT = auto()
    RETURN = auto()
    SUPER = auto()
    THIS = auto()
    TRUE = auto()
    VAR = auto()
    WHILE = auto()
    EOF = auto()


class Token:
    kind: TokenKind
    lexeme: str
    literal: Any
    line: int

    def __init__(self, kind: TokenKind, lexeme: str, literal: Any, line: int):
        self.kind = kind
        self.lexeme = lexeme
        self.literal = literal
        self.line = line

    def __str__(self):
        return f"[kind: {self.kind}] [lexeme: '{self.lexeme}'] [value: {self.literal}]"


class Scanner:
    source: str
    tokens: list[Token] = []
    line: int = 1
    start: int = 0
    current: int = 0

    def __init__(self, source: str):
        self.source = source

    def scan_tokens(self) -> list[Token]:
        while not self.is_at_end():
           self.start = self.current 
           self.scan_token()
        
        self.tokens.append(Token(TokenKind.EOF, '', None, self.line))

        return self.tokens

    def scan_token(self) -> None:
        ch = self.advance()

        single_char_tokens_kinds = {
            '(': TokenKind.LEFT_PAREN,
            ')': TokenKind.RIGHT_PAREN,
            '{': TokenKind.LEFT_BRACE,
            '}': TokenKind.RIGHT_BRACE,
            ',': TokenKind.COMMA,
            '.': TokenKind.DOT,
            '-': TokenKind.MINUS,
            '+': TokenKind.PLUS,
            ';': TokenKind.SEMICOLON,
            '*': TokenKind.STAR,
            # This don't work. The table is built only one time. it's not like a
            # switch statement.
            '!': self.if_matches('=', TokenKind.BANG_EQUAL, TokenKind.BANG),
            '=': self.if_matches('=', TokenKind.EQUAL_EQUAL, TokenKind.EQUAL),
            '<': self.if_matches('=', TokenKind.LESS_EQUAL, TokenKind.LESS),
            '>': self.if_matches('=', TokenKind.GREATER_EQUAL, TokenKind.GREATER),
        }

        if kind := single_char_tokens_kinds.get(ch):
            self.add_token(kind)
            return
        
        error(self.line, f"Unexpected character '{ch}'")

    def add_token(self, kind: TokenKind, literal: Any = None) -> None:
        lexeme = self.source[self.start:self.current]
        self.tokens.append(Token(kind, lexeme, literal, self.line))

    def advance(self) -> None:
        ch = self.source[self.current]
        self.current += 1
        return ch

    def if_matches(self, ch: str, v1: TokenKind, v2: TokenKind) -> TokenKind:
        if self.peek() == ch:
            self.current += 1
            return v1
        return v2
    
    def peek(self) -> str | None:
        if self.is_at_end():
            return None
        return self.source[self.current]

    def is_at_end(self) -> bool:
        return self.current >= len(self.source)
    